### PR TITLE
Patch batchID label to pods during rollout

### PR DESCRIPTION
**Pull request by veophi**
Signed-off-by: mingzhou.swx <mingzhou.swx@alibaba-inc.com>

### Ⅰ. Describe what this PR does
Patch batchID label to pods during rollout.

### CODE REVIEW

File: pkg/util/pod_utils.go [hunk 1]
```diff
@@ -57,3 +64,68 @@ func IsConsistentWithRevision(pod *v1.Pod, revision string) bool {
...
+
+	podLister := &v1.PodList{}
+	err = c.List(context.TODO(), podLister, &client.ListOptions{LabelSelector: selector}, utilclient.DisableDeepCopy)
+	if err != nil {
+		return nil, err
+	}
+	pods := make([]*v1.Pod, 0)
+	for i := range podLister.Items {
+		pod := &podLister.Items[i]
+		if !pod.DeletionTimestamp.IsZero() {
```
[the remark this answers is not in the document]
Comment (veophi) [reply pkg/util/pod_utils.go]: done

File: pkg/controller/batchrelease/workloads/statefulset_like_controller.go [hunk 2]
```diff
@@ -244,3 +217,13 @@ func (c *StatefulSetLikeController) countUpdatedReadyPods(updateRevision string)
 	klog.V(3).Infof("BatchRelease(%v) observed %d updatedReadyReplicas")
 	return updatedReadyReplicas, nil
 }
+
+func isStatefulSetUpgradedDone(set *unstructured.Unstructured, canaryReplicasGoal, stableReplicasGoal int32) bool {
+	partition := util.GetStatefulSetPartition(set)
+	if partition <= stableReplicasGoal {
```
[the remark this answers is not in the document]
Comment (veophi) [reply pkg/controller/batchrelease/workloads/statefulset_like_controller.go:224]: we check it in `IsBatchReady` function

File: pkg/util/pod_utils.go [hunk 3]
```diff
@@ -57,3 +64,68 @@ func IsConsistentWithRevision(pod *v1.Pod, revision string) bool {
...
+			continue
+		}
+		pods = append(pods, pod)
+	}
+	return pods, nil
+}
+
+func PatchPodBatchLabel(c client.Client, pods []*v1.Pod, rolloutID string, batchID int32, updateRevision string, canaryGoal int32, logKey types.NamespacedName) (bool, error) {
+	patchedCount := int32(0)
+	for _, pod := range pods {
```
[the remark this answers is not in the document]
Comment (veophi) [reply pkg/util/pod_utils.go:106]: Calculate `patchedCount` is useful to the following logic to solve rollback case:
https://github.com/openkruise/rollouts/blob/9969d50b19572c8e3446c4affbf453b1ff54a2b0/pkg/util/pod_utils.go#L138

File: config/rbac/role.yaml [hunk 4]
```diff
@@ -6,6 +6,14 @@ metadata:
   creationTimestamp: null
   name: manager-role
 rules:
+- apiGroups:
```
Comment (furykerry) [inline config/rbac/role.yaml:9]: can we add helm values for operator CRDs?  Asking for read privilege for all resources is too demanding

File: pkg/controller/batchrelease/workloads/statefulset_like_controller.go [hunk 5]
```diff
@@ -198,49 +191,40 @@ func (c *StatefulSetLikeController) IsBatchReady(canaryReplicasGoal, stableRepli
...
+	return c.pods, err
 }
 
 func (c *StatefulSetLikeController) countUpdatedReadyPods(updateRevision string) (int32, error) {
-	pods, err := c.listOwnedPods()
+	pods, err := c.ListOwnedPods()
 	if err != nil {
 		return 0, err
 	}
+	activePods := util.FilterActivePods(pods)
```
Comment (furykerry) [inline pkg/controller/batchrelease/workloads/statefulset_like_controller.go:211]: can we reuse FilterActivePods in the controller_utils.go

File: pkg/util/pod_utils.go [hunk 6]
```diff
@@ -57,3 +64,94 @@ func IsConsistentWithRevision(pod *v1.Pod, revision string) bool {
...
+	}
+
+	podLister := &v1.PodList{}
+	err = c.List(context.TODO(), podLister, &client.ListOptions{LabelSelector: selector}, utilclient.DisableDeepCopy)
+	if err != nil {
+		return nil, err
+	}
+	pods := make([]*v1.Pod, 0, len(podLister.Items))
+	for i := range podLister.Items {
+		pod := &podLister.Items[i]
```
Comment (furykerry) [inline pkg/util/pod_utils.go:91]: we can safely filter out pods with p.Status.Phase = PodSucceeded and PodFailed

File: pkg/util/workloads_utils.go [hunk 7]
```diff
@@ -153,7 +153,7 @@ func EqualIgnoreHash(template1, template2 *v1.PodTemplateSpec) bool {
 }
 
 func HashReleasePlanBatches(releasePlan *v1alpha1.ReleasePlan) string {
-	by, _ := json.Marshal(releasePlan.Batches)
+	by, _ := json.Marshal(releasePlan)
```
Comment (furykerry) [inline pkg/util/workloads_utils.go:156]: plz update comments in the. releaseplan.status.ObservedReleasePlanHash

File: pkg/util/workloads_utils.go [hunk 8]
```diff
@@ -428,16 +428,26 @@ func parseMetadataFrom(object *unstructured.Unstructured) *metav1.ObjectMeta {
 	return meta
 }
 
-// ParseSelectorFrom can find labelSelector and parse it as selector for unstructured object
-func ParseSelectorFrom(object *unstructured.Unstructured) (labels.Selector, error) {
-	m, found, err := unstructured.NestedFieldNoCopy(object.Object, "spec", "selector")
-	if err != nil || !found {
-		return nil, err
+// ParseSelector can find labelSelector and parse it as labels.Selector for client object
+func ParseSelector(object client.Object) (labels.Selector, error) {
```
Comment (furykerry) [inline pkg/util/workloads_utils.go:432]: no need to publish this func

File: pkg/util/pod_utils.go (hunk 6 above)
Comment (veophi) [reply pkg/util/pod_utils.go:95]: done

File: pkg/util/workloads_utils.go (hunk 8 above)
Comment (veophi) [reply pkg/util/workloads_utils.go:432]: done

File: pkg/util/workloads_utils.go (hunk 7 above)
Comment (veophi) [reply pkg/util/workloads_utils.go:156]: done

File: pkg/controller/batchrelease/workloads/statefulset_like_controller.go (hunk 5 above)
Comment (veophi) [reply pkg/controller/batchrelease/workloads/statefulset_like_controller.go:211]: It may  introduce too many dependencies.

**Review by furykerry — COMMENTED** [review]
/lgtm

**Pull request merged**